Implement `Client#send_raw_request`

jugglinmike opened the pull request:
Introduce a new method to allow tests to explicitly inspect the status
code and complete body of HTTP responses received when issuing commands.
In addition, extend the behavior of the higher-order `send_response`
method with assertions for HTTP response headers.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/wdclient/2)
<!-- Reviewable:end -->

jugglinmike committed:
Implement `Client#send_raw_request`

Introduce a new method to allow tests to explicitly inspect the status
code and complete body of HTTP responses received when issuing commands.
In addition, extend the behavior of the higher-order `send_response`
method with assertions for HTTP response headers.

jugglinmike committed:
Validate all command responses

jugglinmike committed:
Correct string representation

jugglinmike committed:
Rename and re-locate `Result` class

andreastt commented on webdriver/result.py:
```diff
@@ -0,0 +1,31 @@
+import json
+
+class Result():
```
Extend `object` and put it in webdriver/transport.py.

I also think you should consider renaming this to `Response` to match the spec language. A “result” seems more like a generic type.

andreastt commented on webdriver/result.py:
```diff
@@ -0,0 +1,31 @@
+        self._validate(response)
+
+    def __repr__(self):
+        return 'wdclient.Result(status=%d, body=%s)' % (self.status, self.body)
```
Use double quotes throughout for consistency.

jugglinmike replied:
Uhg, that was sloppy. Will do.

andreastt commented on webdriver/result.py:
```diff
@@ -0,0 +1,31 @@
+        # >    "cache-control"
+        # >       "no-cache"
+        assert response.getheader("Content-Type") == "application/json; charset=utf-8"
+        assert response.getheader("cache-control") == "no-cache"
```
The spec has since been fixed and this should say `Cache-Control` now.

jugglinmike replied:
Character case doesn't actually matter for HTTP headers, and [the standard library accounts for
this](https://hg.python.org/cpython/file/2.7/Lib/rfc822.py#l297). I was on the fence about using the idiomatic spelling versus mirroring the spec, so I'm glad you made them one and the same!

andreastt replied:
OK, that’s good to know.

andreastt commented on webdriver/result.py:
```diff
@@ -0,0 +1,31 @@
+
+        # > 4. If data is not null, let response's body be a JSON Object with a
+        #      key value set to the JSON Serialization of data.
+        if self.body:
```
Use `if self.body is not None` because we enter this branch if the value is falsy (0, False, "", &c.).

andreastt commented on webdriver/transport.py:
```diff
@@ -33,15 +31,14 @@ def __init__(self, host, port, url_prefix="/", timeout=HTTP_TIMEOUT):
         :param headers: Additional headers to include in the request.
-        :param key: Extract this key from the dictionary returned from
-            the remote.
         """
```
Please document what this returns now that it returns `Result`.

andreastt commented on webdriver/result.py:
```diff
@@ -0,0 +1,31 @@
+        if self.body:
+            self.body = json.loads(self.body)
+
+        self._validate(response)
```
I dislike constructors that do too much processing as it weakens the trust in that an object can be constructed without adverse consequences. Can we create a static method on this object, `from_response` (or equivalent) that reads, parses, and validates it? Then it can pass the valid data pieces into this ctor.

andreastt commented on webdriver/client.py:
```diff
@@ -236,11 +237,33 @@ def end(self):
+        url = urlparse.urljoin("session/%s/" % self.session_id, url)
+        return self.transport.send(method, url, body, headers)
+
     def send_command(self, method, url, body=None, key=None):
```
We should at some point remove `key` from here too, but probably not in this PR.

jugglinmike replied:
I'll document it as "(deprecated)" for now

andreastt commented on webdriver/client.py:
```diff
@@ -236,11 +237,33 @@ def end(self):
+        :param headers: Additional headers to include in the HTTP request
+        """
+        url = urlparse.urljoin("session/%s/" % self.session_id, url)
+        return self.transport.send(method, url, body, headers)
```
Document return value.

andreastt commented on webdriver/client.py:
```diff
@@ -236,11 +237,33 @@ def end(self):
+        url = urlparse.urljoin("session/%s/" % self.session_id, url)
+        return self.transport.send(method, url, body, headers)
+
     def send_command(self, method, url, body=None, key=None):
```
Please provide docstring for this too.

jugglinmike replied:
Sure thing

jugglinmike committed:
Extend documentation

jugglinmike committed:
Correct quotation usage

jugglinmike reviewed (commented):
Thanks for the review, Andres! The factoring you've requested is much better. I think I've addressed all your requests

jugglinmike commented:
@andreastt I think I've addressed all your feedback, although the "Reviewable" application may not have recognized that. Please let me know if there's anything more I can do here :)

andreastt commented:
Merged in 6eb2b1484f90aa783c615bf113c5bc3178423e3d...97fcf19f9903ddb0d2d2564a6ef8d5320559c19d.
